### DISCUSSION TITLE
Improve zip documentation

### DISCUSSION THREAD

**Discussion by HarelM** [101arrowz/fflate#185] [Q&A] [answered]
**What can't you do right now?**

I can't understand how to create a zip file and how to read one with folders etc.

**An optimal solution**

Improve the docs.

**(How) is this done by other libraries?**

I'm using jszip where you can easily add a string and create a file, and browse directories.

I'm also using pako to gunzip, and I would like to use a single library for all my compression and decompression work.
Currently the docs provided here for adding a zip file is partial - what is "chunk1" and "chunk2", how do I iterate over a folder etc...

Here are some of the examples I would like to migrate to this library:
https://github.com/IsraelHikingMap/Site/blob/ba1c6a10ac40d6e429bbeb15778fef8479800a84/IsraelHiking.Web/src/application/services/file.service.ts#L156

https://github.com/IsraelHikingMap/Site/blob/ba1c6a10ac40d6e429bbeb15778fef8479800a84/IsraelHiking.Web/src/application/services/file.service.ts#L245

https://github.com/IsraelHikingMap/Site/blob/ba1c6a10ac40d6e429bbeb15778fef8479800a84/IsraelHiking.Web/src/application/services/file.service.ts#L257

Any help would be appreciated.

**Comment by 101arrowz** [comment, marked as answer]
To zip a file (like in your first link):

```js
const result = fflate.zipSync({ 'log.txt': fflate.strToU8('some text') });

// convert to a blob if you need
const resultBlob = new Blob([result]);
```

To unzip (like in your other link):

```js
const zipData = new Uint8Array(await yourBlob.arrayBuffer());
const files = fflate.unzipSync(zipData, {
  filter: file => file.name.startsWith("styles/") && file.name.endsWith(".json")
});

for (const fileName in files) {
  const styleText = fflate.strFromU8(files[fileName]);
  // do whatever you want here
}
```

If you care about not blocking the main thread, you can replace the `zipSync` or `unzipSync` calls with `zip` or `unzip`, e.g.

```js
// first example
const result = await new Promise((resolve, reject) => fflate.zip(
   { 'log.txt': fflate.strToU8('some text') },
  (err, result) => err ? reject(err) : resolve(result)
);

// second example
const files = await new Promise((resolve, reject) => fflate.unzip(zipData, {
  filter: file => file.name.startsWith("styles/") && file.name.endsWith(".json")
}, (err, result) => err ? reject(err) : resolve(result)));
```

For reference, the ZIP documentation is quite thorough; you may have been looking in the wrong place.
<details>
<summary>From the README (click to expand)</summary>

```js
// Note that the asynchronous version (see below) runs in parallel and
// is *much* (up to 3x) faster for larger archives.
const zipped = fflate.zipSync({
  // Directories can be nested structures, as in an actual filesystem
  'dir1': {
    'nested': {
      // You can use Unicode in filenames
      '你好.txt': fflate.strToU8('Hey there!')
    },
    // You can also manually write out a directory path
    'other/tmp.txt': new Uint8Array([97, 98, 99, 100])
  },

  // You can also provide compression options
  'massiveImage.bmp': [aMassiveFile, {
    level: 9,
    mem: 12
  }],
  // PNG is pre-compressed; no need to waste time
  'superTinyFile.png': [aPNGFile, { level: 0 }],

  // Directories take options too
  'exec': [{
    'hello.sh': [fflate.strToU8('echo hello world'), {
      // ZIP only: Set the operating system to Unix
      os: 3,
      // ZIP only: Make this file executable on Unix
      attrs: 0o755 << 16
    }]
  }, {
    // ZIP and GZIP support mtime (defaults to current time)
    mtime: new Date('10/20/2020')
  }]
}, {
  // These options are the defaults for all files, but file-specific
  // options take precedence.
  level: 1,
  // Obfuscate last modified time by default 
  mtime: new Date('1/1/1980')
});

// If you write the zipped data to myzip.zip and unzip, the folder
// structure will be outputted as:

// myzip.zip (original file)
// dir1
// |-> nested
// |   |-> 你好.txt
// |-> other
// |   |-> tmp.txt
// massiveImage.bmp
// superTinyFile.png

// When decompressing, folders are not nested; all filepaths are fully
// written out in the keys. For example, the return value may be:
// { 'nested/directory/structure.txt': Uint8Array(2) [97, 97] }
const decompressed = fflate.unzipSync(zipped, {
  // You may optionally supply a filter for files. By default, all files in a
  // ZIP archive are extracted, but a filter can save resources by telling
  // the library not to decompress certain files
  filter(file) {
    // Don't decompress the massive image or any files larger than 10 MiB
    return file.name != 'massiveImage.bmp' && file.originalSize <= 10_000_000;
  }
});
```
</details>

**Comment by HarelM** [comment]
I will share my experience, let me know how I can help.
I found a mention of this library somewhere and saw that in the readme it talks about the fact that it can handle both gzip and zip, and with very good performance, I got intrigued since I have both and need good performance :-)
I read the readme a few times but couldn't get the exact details.
I then tried to copy the examples from the readme and saw that they are incomplete.
I looked for a link in the readme to a docs website, but I couldn't find one, but I did found a link to an internal readme in this repo.
I clicked on it and got to:
https://github.com/101arrowz/fflate/blob/master/docs/README.md
Which is good in terms of API documentation, but it lacks some intro or a guide of how to read the docs.
I then tried to click on the zip class and got to:
https://github.com/101arrowz/fflate/blob/master/docs/classes/Zip.md
Which is a 404 not found page.
After this experience I decided the docs are probably not maintained and went to look at the tests, which I saw are in "TODO test...".
After this I opened this issue...

As a side note, I recently made the following docs site:
https://maplibre.org/maplibre-gl-js/docs/
It uses mkdocs material and typedoc which are great for making good documentation, I highly recommend them!
The most valuable part of the documentation site I created is the examples - they are giving a short snippet of how to use the code easily.
I'm sure the above-mentioned information is great, but if this is hard to find it kinda misses the point.

Thanks for the super-fast response and the detailed answer, it's very much appreciated!

**Reply by 101arrowz** [reply]
The docs links not working is an issue - I think there's some bug with the casing unfortunately. I'll see if I can fix that. The tests not being fully complete are a very long-standing issue (#3); admittedly I really should have gotten to that by now, the tests handle the most common use cases but not every possible one.

That's a very nicely designed documentation site; I'm already using Typedoc so it might be possible to create something like that without too much trouble. Thanks for the feedback!

**Reply by 101arrowz** [reply]
I've updated the generated documentation; hopefully it's more usable now that the links actually work. I'll still see how doable making a proper docs site with more examples would be.

**Reply by HarelM** [reply]
Here's my typedocs config file:
https://github.com/maplibre/maplibre-gl-js/blob/be14c414d896ca0c3fd71d37c93bc65a157fdedb/typedoc.json
My mkdocs config:
https://github.com/maplibre/maplibre-gl-js/blob/be14c414d896ca0c3fd71d37c93bc65a157fdedb/mkdocs.yml
And the generation script:
https://github.com/maplibre/maplibre-gl-js/blob/be14c414d896ca0c3fd71d37c93bc65a157fdedb/package.json#L148

It uploads the generated site to gh pages on every release:
https://github.com/maplibre/maplibre-gl-js/blob/be14c414d896ca0c3fd71d37c93bc65a157fdedb/.github/workflows/release.yml#L129
